src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala

```diff
@@ -2,49 +2,57 @@ package mesosphere.marathon
 package core.task.update.impl.steps
 //scalastyle:off
 import javax.inject.Named
-
 import akka.Done
 import akka.actor.ActorRef
 import com.google.inject.{Inject, Provider}
 import com.typesafe.scalalogging.StrictLogging
 import mesosphere.marathon.MarathonSchedulerActor.ScaleRunSpec
 import mesosphere.marathon.core.condition.Condition
+import mesosphere.marathon.core.instance.Instance
 import mesosphere.marathon.core.instance.update.{InstanceChange, InstanceChangeHandler}
 
 import scala.concurrent.Future
 //scalastyle:on
 /**
-  * Trigger rescale of affected app if a task died or a reserved task timed out.
+  * This logic covers part of our orchestration logic and will ask to check the
+  * scale of a service when instances become UnreachableInactive, or come back
+  * from that state into one that is considered active.
+  *
+  * The result will be a ScaleRunSpec directive. This will be ignored during
+  * ongoing deployments for the service. In this case, the responsible
+  * deployment logic will have to handle these cases.
   */
 class ScaleAppUpdateStepImpl @Inject() (
     @Named("schedulerActor") schedulerActorProvider: Provider[ActorRef]) extends InstanceChangeHandler with StrictLogging {
 
   private[this] lazy val schedulerActor = schedulerActorProvider.get()
 
-  private[this] def scalingWorthy: Condition => Boolean = {
-    case Condition.Reserved | Condition.UnreachableInactive | _: Condition.Terminal => true
-    case _ => false
-  }
-
   override def name: String = "scaleApp"
   override def metricName: String = "scale-app"
 
   override def process(update: InstanceChange): Future[Done] = {
-    // TODO(PODS): it should be up to a tbd TaskUnreachableBehavior how to handle Unreachable
-    calcScaleEvent(update).foreach(event => schedulerActor ! event)
+    // This doesn't block on the result of a scale check, which means that
+    // fast subsequent instance changes for the same PathId will eventually be
+    // ignored, since the pathId might be locked from a previous ScaleRunSpec
+    // attempt. This happened in tests, so it should be considered a real problem.
+    // 1. We would send an event to the schedulerActor
+    // 2. The schedulerActor would lock the app and call schedulerActions.scale
+    // 3. If one or more InstanceChange are processed before the lock is released, no replacements are scheduled for those
+    // The logs say e.g. `Did not try to scale run spec /app-kill-all-tasks-of-an-app; it is locked`
+    // An option would be to block here until the scale request is acked.
+    update.lastState.foreach { lastState =>
```

Comment:
update.lastState is introduced so we will ignore new instances? The reason for this is unclear. My reading of the code tells me that before, calcScaleEvent could return a `Some` for instances without a lastState, due do the fact that this always returns true when update.lastState is None:

```scala
update.lastState.forall(lastState => !scalingWorthy(lastState.condition))
```

Can you elaborate more on this? Why do we prevent a ScaleRunSpec now from being sent, where before it was possible? Was it an accident that new instances could trigger a ScaleRunSpec before?

Reply:
yes, right now we scale check only on a transition between `Unreachable` and `UnreachableInactive` so we can omit the states without lastState (that's only in case of reserve or schedule)